design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,276 @@
+
+A `Service Config` is a JSON object that specifies information required for a service to be created e.g. for Conjur `CONJUR_ACCOUNT`, `version` etc. Conjur will have different configuration options to Postgres. These options must be documented for each Service. Ideally this would be in a centralized repository. It's important for all users of a Service to know what they're able to configure on a service.
+```
+{
+  "conjur_account": "test",
+  "version": "10.9",
+  "admin_user": "kumbi",
+  "admin_password": "definetelyasecret"
+}
+```
+
+A `Service Definition` is a JSON object that specifies the lifecyle methods of a Service (start, stop). 
+    1. `start_command`. This command sets up the service and outputs any meaningful information as part of the JSON output. The JSON output must be written to the named pipe which is passed in as the second argument to the command. The first argument is the Service Configuration. The JSON output of any given Service must have a well-defined and documented schema.
+    2. `stop_command`. This command carries out any required regardless of test outcome. It is important for the test runner to ensure that this is run on both success and failure
+    3. `service_type`. This is optional and for the future. The `service_type` specifies a internal implementation of a Service. The internal implementation come from a standard library provided by and maintened by the Infrastructure team.
+```
+{
+  "service_name": "mysql_service",
+  "service_type": "myql",
+  "start_command": "mysql_start.sh",
+  "stop_command": "mysql_stop.sh"
+}
+```
+
+A `Test Runner`  is software that provides a conventional way for defining and running test cases. They typically come with capabilities to measure test metrics, test outcomes and code coverage. This will also allow historical tracking. An example of a test-runner is Cucumber. Test cases can be defined as a sequence of steps in the Gherkin language. The step definitions are written in the language of the test runner e.g. Python, Ruby, Go. In this case we're strongly considering adopting `Cucumber` in Ruby as the test-runner. Ruby because of the ubiquity of working knowledge of that language in the team and the fact that there are Ruby Cucumber test suites that exist within some of our repos.
+
+The `Service Engine` is a combination of steps defined in the `Test Runner`. The Engine orchestrates the lifecycle of a Service. It provides step definitions for starting Services, stopping services regardless of the test outcome, creating named pipes for the Service output and writing that to the test context.
+
+A `Test Case` is a sequence of steps that goes through a scenario to validate some given functianlity. It should really include at least one step making an assertion as part of the validation.
+
+A `Context` is a temporary store within a test case. Steps can read and write from the store. The key `service_name` will be used to identify this output.
```

Comment:
In general we should try to minimize non-idempotent steps (and thus use of Context)